databricks/koalas/frame.py

```diff
@@ -7237,46 +7260,109 @@ def sample(
+                weights = weights.reindex(self.axes[axis])
+
+            # Strings acceptable if a dataframe and axis = 0
+            if isinstance(weights, str):
+                if isinstance(self, ks.DataFrame):
+                    if axis == 0:
+                        try:
+                            weights = self[weights]
+                        except KeyError as err:
+                            raise KeyError("String passed to weights not a valid column") from err
+
+            # Because ks.Series currently does not support the Series.__iter__ method,
+            # It cannot be initialized to the pandas Series, so here is to_pandas.
+            if isinstance(weights, ks.Series):
+                weights = pd.Series(weights.to_pandas(), dtype="float64")
+            else:
+                weights = pd.Series(weights, dtype="float64")
+
+            if len(weights) != axis_length:
+                raise ValueError("Weights and axis to be sampled must be of same length")
+
+            if (weights == np.inf).any() or (weights == -np.inf).any():
+                raise ValueError("weight vector may not include `inf` values")
+
+            if (weights < 0).any():
+                raise ValueError("weight vector many not include negative values")
+
+            # If has nan, set to zero.
+            weights = weights.fillna(0)
+
+            # Renormalize if don't sum to 1
+            if weights.sum() != 1:
+                if weights.sum() != 0:
+                    weights = weights / weights.sum()
+                else:
+                    raise ValueError("Invalid weights: weights sum to zero")
+
+            weights = weights._values
+
+        # If no frac or n, default to n=1.
+        if n is None and frac is None:
+            n = 1
+        elif frac is not None and frac > 1 and not replace:
+            raise ValueError(
+                "Replace has to be set to `True` when " "upsampling the population `frac` > 1."
+            )
+        elif n is not None and frac is None and n % 1 != 0:
+            raise ValueError("Only integers accepted as `n` values")
+        elif n is None and frac is not None:
+            n = int(round(frac * axis_length))
+        elif n is not None and frac is not None:
+            raise ValueError("Please enter a value for `frac` OR `n`, not both")
+
+        # Check for negative sizes
+        if n < 0:
+            raise ValueError("A negative number of rows requested. Please provide positive value.")
+
+        # Because duplicated row selection is not currently supported.
+        # So if frac > 1, use the pyspark implementation.
+        if frac is not None and frac > 1:
+            sdf = self._internal.resolved_copy.spark_frame.sample(
+                withReplacement=replace, fraction=float(frac), seed=random_state
+            )
+            return DataFrame(self._internal.with_new_sdf(sdf))
+        locs = rs.choice(axis_length, size=n, replace=replace, p=weights)
```

Comment:
What's the estimated size of `locs`? Will it also be much huge?
e.g., for the case `A random 50% sample of the ``DataFrame`` with replacement`?

Reply:
The `rs.choice` method is similar to the [numpy.random.RandomState.choice](https://numpy.org/doc/stable/reference/random/generated/numpy.random.RandomState.choice.html).
The return value of `rs.choice` is the size of the `size` parameter, in this case is `n`.
So the size of `locs` here is `n`.
If parameter `n` is set, `locs` is equal to `n`, otherwise `locs` is equal to `n = int(round(frac * axis_length))`.
For the case `A random 50% sample of the ``DataFrame`` with replacement`, the size of `locs` will be one-half of the size of the `DataFrame`, to be precise, it should be `int(round(0.5 * DataFrame.nrows))`.
If `n` is large, or `frac` is large, `locs` will be large.
Performance depends on the performance of the `take` method, which is actually the performance of the `iloc` method

Reply:
In that case, I wouldn't recommend to use `take` or `iloc` for this.
Thinking of Koalas workload, the length of `DataFrame` could be so huge, then `locs` will be too huge for a single Driver node.
Also, row access by its row number is essentially heavy on Spark, and so `iloc` is heavy in Koalas, especially if the `locs` is huge.

Reply:
Yeah!
I agree with you, because spark does not have the concept of rowindex, iloc is a heavy operation. 
But here the weights parameter of the sample function is the same as iloc, which must depend on the row index. 
There may be no other way to support the weights parameter right now. 
Just like iloc is a heavy operation, but it must also be implemented based on row index right now. 
Maybe the current sample function supports weights operation and must also be based on row index